druid/src/core.rs

```diff
@@ -473,17 +473,29 @@ impl<T: Data, W: Widget<T>> WidgetPod<T, W> {
                 r
             }
             LifeCycle::WidgetAdded => {
-                // if this is called, it means widgets were added. That might be
-                // be us; if we don't have old_data we set it now.
+                assert!(self.old_data.is_none());
+
+                self.old_data = Some(data.clone());
+                self.env = Some(env.clone());
+
+                true
+            }
+            LifeCycle::RouteWidgetAdded => {
+                // if this is called either we were just created, in
+                // which case we need to change lifecycle event to
+                // WidgetAdded or in case we were already created
+                // we just pass this event down
                 if self.old_data.is_none() {
-                    self.old_data = Some(data.clone());
-                    self.env = Some(env.clone());
-                }
-                if self.state.children_changed {
-                    self.state.children.clear();
-                    self.state.focus_chain.clear();
+                    self.lifecycle(ctx, &LifeCycle::WidgetAdded, data, env);
+                    return;
```

Comment:
oh, this is confusing? If we return here then we aren't going to end up registering ourselves, which normally happens at the end of this fn (where we call `ctx.register_child(self.id())`).

It's possible that we don't have a test case to catch this?

let me take a look and see, it's worth having if we don't.

Reply:
okay, I've opened #606 and I'm going to merge that shortly; when that's merged you can rebase this and then if that test is passing I'll be happy. :)

Reply:
I rebased on top of master!

Since we enter the function again recursively with a different argument (WidgetAdded instead of RouteWidgetAdded) we still end up at the end of the function for 'self'. So we end up registering ourselves I think.

Reply:
ah okay I didn't catch that, I read this as a call to the child's lifecycle fn.